pkg/controller/registry/reconciler/grpc.go

```diff
@@ -185,57 +186,51 @@ func (c *GrpcRegistryReconciler) ensureUpdatePod(source grpcCatalogSourceDecorat
 	if !source.Poll() {
 		return nil
 	}
-	var updateFlag bool
-	currentLivePods := c.currentPods(source)
-	currentUpdatePods := c.currentUpdatePods(source)
-
-	for _, updatePod := range currentUpdatePods {
-		if updatePodByDigest(updatePod, currentLivePods) {
-			logrus.WithField("CatalogSource", source.GetName()).Info("detect image update for catalogsource pod")
-
-			updateFlag = true
-			updatePod.Labels[CatalogSourceLabelKey] = source.GetName()
-			updatePod.Labels[CatalogSourceUpdateKey] = ""
 
-			// Update the update pod to promote it to serving pod
-			_, err := c.OpClient.KubernetesInterface().CoreV1().Pods(source.GetNamespace()).Update(context.TODO(), updatePod, metav1.UpdateOptions{})
-			if err != nil {
-				return errors.Wrapf(err, "error creating new pod: %s", source.Pod().GetName())
-			}
+	if !source.Update() {
+		logrus.WithField("CatalogSource", source.GetName()).Infof("no catalog update required at %#v", time.Now())
+		return nil
+	}
 
-			break
-		}
+	currentLivePods := c.currentPods(source)
+	updatePod, err := c.createUpdatePod(source)
+	if err != nil {
+		return errors.Wrapf(err, "creating update catalog source pod")
 	}
+	source.SetLastUpdateTime()
 
-	if updateFlag {
-		// Clean up outdated serving pod(s)
-		err := c.removePods(currentLivePods, source.GetNamespace())
+	// TODO bail out at some point if the pod never becomes ready
+	for {
```

Comment:
We shouldn't spin like this in the sync loop -  

Instead we should check for the exit condition we want (update pod ready) and if it's there, do the rest of the work, and if not, requeue the current object to check again in a bit. This keeps the controller running and processing other objects

Reply:
In general, this is the pattern we should use for all eventually consistent work. Since the queue that feeds the reconciliation has a constant throughput (e.g. GOMAXPROC), blocking any one sync can prevent events on other resources from being reconciled. In this case, we could prevent other CatalogSources from being noticed until a pod becomes ready. Even with a timeout on this `for`, we would see that latency reflected down the queue.

Reply:
Ok I think I addressed this in the latest commit by adding a requeue based on the type of error the `ensureUpdatePod()` function returns. 

The only concern I have is if the polling interval is extremely small this code will always spin up a new update pod. That needs to be fixed. But I think its close.

Reply:
I fixed this by adding a `&& len(currentUpdatePods) == 0` to the creation logic. But this basically makes those really small polling intervals not work